### PR TITLE
Confiant RTD Module : initial release

### PR DESCRIPTION
**Pull request by denys-berzoy-confiant**
## Type of change
- [x] Feature
- [x] Does this change affect user-facing APIs or examples documented on http://prebid.org?

## Description of change
Added Confiant RTD Provider Module

## Other information
Relevant docs PR: https://github.com/prebid/prebid.github.io/pull/4202

### CODE REVIEW

**Review by garygz-confiant — APPROVED** [review]

**Comment by garygz-confiant** [comment]
@denys-berzoy-confiant looks like you need to pull the master - see conflicts above

File: modules/confiantRtdProvider.js [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    });
+  });
+  mutationObserver.observe(document.head, { childList: true, subtree: true });
+}
+
+/**
+ * Emit billable event when Confiant integration reports that it has monitored an impression
+ */
+function reportBillableEvents (e) {
+  events.emit(CONSTANTS.EVENTS.BILLABLE_EVENT, {
```
Comment (patmmccann) [inline modules/confiantRtdProvider.js:88]: why bother? doesnt your js already do this emit?

File: modules/confiantRtdProvider.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+/**
+ * Injects the Confiant Inc. configuration script into the page, based on proprtyId provided
+ * @param {string} propertyId
+ */
+function injectConfigScript(propertyId) {
+  const scriptElement = document.createElement('script');
+  scriptElement.type = 'text/javascript';
+  scriptElement.src = `https://cdn.confiant-integrations.net/${propertyId}/gpt_and_prebid/config.js`;
+
+  insertElement(scriptElement);
```
Comment (patmmccann) [inline modules/confiantRtdProvider.js:25]: you must use our methods, please import the ajax utility

**Review by patmmccann — CHANGES_REQUESTED** [review]
couple of minor comments in line

File: modules/confiantRtdProvider.js (hunk 1 above)
Comment (denys-berzoy-confiant) [reply modules/confiantRtdProvider.js:88]: @patmmccann no, there we no use cases to do that from our codebase.

File: modules/confiantRtdProvider.js (hunk 2 above)
Comment (denys-berzoy-confiant) [reply modules/confiantRtdProvider.js:25]: @patmmccann this is pretty much the same approach that Clean.io is taking to inject their script. Do you mean we should use ajax to fetch .js and the `eval()` it?

Comment (garygz-confiant) [reply modules/confiantRtdProvider.js:25]: @patmmccann just like Clean.io our client asked to implement this to cover other demand as well native, GPT as well if they choose to do so.
So we are just replicating the approach that was approved for the Clean.io prebid module, see the link above.

**Comment by garygz-confiant** [comment]
@robertrmartinez see this Confiant PR

File: modules/confiantRtdProvider.js (hunk 1 above)
Comment (patmmccann) [reply modules/confiantRtdProvider.js:88]: We're a confiant customer, and you have no trouble billing is, what value is this event adding

File: modules/confiantRtdProvider.js (hunk 2 above)
Comment (patmmccann) [reply modules/confiantRtdProvider.js:25]: They are not similar and the change request stands.

Comment (patmmccann) [reply modules/confiantRtdProvider.js:25]: I looked at this again; the reason I flagged them as not similar is bc you linked the wrong part of the clean adapter, preloading is not against the rules (although we are considering it). The correct part of the clean adapter that inserts has been added to this list https://github.com/prebid/Prebid.js/issues/8421

Comment (garygz-confiant) [reply modules/confiantRtdProvider.js:25]: @denys-berzoy-confiant it seems that the new way of inserting scripts is this: https://github.com/prebid/Prebid.js/pull/8496/files#diff-daa702a1f45a1ffc6ff61f98bfd8b733621e125a5525b3c2927e53294f2a0fd5R14
please update the PR

Comment (patmmccann) [reply modules/confiantRtdProvider.js]: this is resolved now ty

**Comment by garygz-confiant** [comment]
> @garygz-confiant only outstanding thing is explanation of how the billing event is used. This pr largely seems to be using prebid as a tag manager and adding this billing event.
> 
> Should your existing customers be switching to this implementation so they can be billed more accurately? If not, why fire the event? Is the event optional? Please make this clear in the module doc.

@patmmccann we are developing this for Magnite who explicitly requested the billable event to be communicated to them in a standard way. All of their vendors communicate this event in a unified way. While we have our own proprietary event (aka pixel) and can bill them based on that, that's not how they prefer to operate.

This is event is optional and configurable by a feature flag on our side (see this flag setup: https://github.com/prebid/Prebid.js/pull/9325/files#diff-d09c33d6d7490d64e1725f7700d943fcb2875f189a561eebf5506c5381ec4102R41). We would not require our existing customers to switch to processing this event as you correctly noticed our proprietary pixels are adequate. 

I see your point in terms of making it more explicitly parameterized and asking Magnite to initilize it while keeping the default as false for other potential clients who choose not to enable it and rather rely on our own billable reports.

@denys-berzoy-confiant Can you change the logic in here:
https://github.com/prebid/Prebid.js/pull/9325/files#diff-d09c33d6d7490d64e1725f7700d943fcb2875f189a561eebf5506c5381ec4102R41

Add an explicit flag:
shouldEmitBillableEvent: true|false
and use it in https://github.com/prebid/Prebid.js/pull/9325/files#diff-d09c33d6d7490d64e1725f7700d943fcb2875f189a561eebf5506c5381ec4102R41

By default it should be false. We will ask Magnite to explicitly enable it.

**Comment by patmmccann** [comment]
@garygz-confiant thanks so much, please document your comments in your pr and i will merge

https://github.com/prebid/prebid.github.io/pull/4202/files

**Comment by garygz-confiant** [comment]
> @garygz-confiant thanks so much, please document your comments in your docs pr, change that default as you indicated, and i will merge; next release is tomorrow so you should be good
> 
> https://github.com/prebid/prebid.github.io/pull/4202/files

@patmmccann We have adjusted the code as per our discussion. The billable event is off by default and we'll be enabled only by clients who explicitly require it:
https://github.com/prebid/Prebid.js/pull/9325/files#diff-01e26e1ce25e326e331c4ddf23dd08846e4fa2cb9234b78c11f5057d6b298014R40

**Comment by patmmccann** [comment]
Lgtm

**Review by patmmccann — APPROVED** [review]

File: modules/confiantRtdProvider.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
+/**
+ * This module provides comprehensive detection of security, quality, and privacy threats by Confiant Inc,
+ * the industry leader in real-time detecting and blocking of bad ads
+ *
+ * The {@link module:modules/realTimeData} module is required
+ * The module will inject a Confiant Inc. script into the page to monitor ad impressions
+ * @module modules/cleanioRtdProvider
```
Comment (muuki88) [inline modules/confiantRtdProvider.js:7]: @denys-berzoy-confiant I guess this should be confiant 😅

**Pull request merged**